src/control/server/storage/scm/firmware.go

```diff
@@ -122,6 +133,59 @@ func (p *Provider) QueryFirmware(req FirmwareQueryRequest) (*FirmwareQueryRespon
 	return resp, nil
 }
 
+func (p *Provider) getRequestedModules(requestedUIDs []string, ignoreMissing bool) (storage.ScmModules, error) {
+	modules, err := p.backend.Discover()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(requestedUIDs) == 0 {
+		return modules, nil
+	}
+
+	uniqueUIDs := common.DedupeStringSlice(requestedUIDs)
+	sort.Strings(uniqueUIDs)
+
+	result := make(storage.ScmModules, 0, len(modules))
+	for _, uid := range uniqueUIDs {
+		mod, err := getModule(uid, modules)
+		if err != nil {
+			if ignoreMissing {
+				continue
+			}
+			return nil, err
+		}
+		result = append(result, mod)
+	}
+
+	return result, nil
+}
+
+func getModule(uid string, modules storage.ScmModules) (*storage.ScmModule, error) {
+	for _, mod := range modules {
+		if mod.UID == uid {
+			return mod, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no module found with UID %q", uid)
+}
+
+func filterModules(modules storage.ScmModules, fwRev string, modelID string) storage.ScmModules {
+	filtered := make(storage.ScmModules, 0, len(modules))
+	for _, mod := range modules {
+		if filterMatches(fwRev, mod.FirmwareRevision) &&
+			filterMatches(modelID, mod.PartNumber) {
+			filtered = append(filtered, mod)
+		}
+	}
+	return filtered
+}
+
+func filterMatches(filterStr, actualStr string) bool {
```

Comment:
This looks a lot like the same function in src/control/server/storage/bdev/firmware.go.  Maybe put it into a package in common between them.

Reply:
Good idea. Let me see where we could share this.

Reply:
Fixed